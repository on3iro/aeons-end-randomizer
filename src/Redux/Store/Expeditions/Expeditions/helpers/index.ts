export * from './convertExpeditionToConfig'
export * from './convertExpeditionFromConfig'

export const calcBattleScore = (tries: number) => {
  switch (tries) {
    case 1: {
      return 6
    }

    case 2: {
      return 4
    }

    case 3: {
      return 2
    }

    default: {
      return 0
    }
  }
}
