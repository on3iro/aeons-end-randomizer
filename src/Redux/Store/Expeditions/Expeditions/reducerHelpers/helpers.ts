import { loop, Cmd } from 'redux-loop'
import { set as setToDb } from 'idb-keyval'
import * as types from 'aer-types/types'

import { State } from '../types'
import { actions } from '../actions'

export const EXPEDITIONS_DB_KEY = 'expeditions-1.11.0'

export const updateBattle = (
  state: State,
  battle: types.Battle,
  additionalBattleProps?: {
    status?: types.BattleStatus
    tries?: number
    rewards?: types.Rewards
  }
) => {
  const oldExpedition = state.expeditions[battle.expeditionId]
  const branches = oldExpedition.sequence.branches

  const updatedBranches = {
    ...branches,
    [battle.id]: {
      ...battle,
      ...additionalBattleProps,
    },
  }

  const newState = {
    ...state,
    expeditions: {
      ...state.expeditions,
      [battle.expeditionId]: {
        ...oldExpedition,
        sequence: {
          ...oldExpedition.sequence,
          branches: updatedBranches,
        },
      },
    },
  }

  return loop(
    newState,
    Cmd.run(setToDb, {
      args: [EXPEDITIONS_DB_KEY, newState],
      successActionCreator: actions.setToDBSuccessful,
      failActionCreator: actions.setToDBFailed,
    })
  )
}
