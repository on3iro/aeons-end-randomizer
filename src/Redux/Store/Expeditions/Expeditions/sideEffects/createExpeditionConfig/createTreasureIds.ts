import {
  createArrayWithDefaultValues,
  createIdList,
  getRandomEntity,
} from 'Redux/helpers'
import * as types from 'types'

export const createTreasureIds = (
  firstBattleConfig: types.BattleConfig,
  availableLevel1TreasureIds: string[],
  seed: types.Seed
) => {
  const startsWithTreasure =
    firstBattleConfig.tier > 1 && firstBattleConfig.treasure.hasTreasure

  return startsWithTreasure
    ? createIdList(
        availableLevel1TreasureIds,
        createArrayWithDefaultValues(5, 'EMPTY'),
        getRandomEntity,
        seed
      )
    : { result: [], seed }
}
