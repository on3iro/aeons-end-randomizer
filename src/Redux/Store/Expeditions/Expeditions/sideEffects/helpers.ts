import * as types from 'aer-types/types'

import {
  createArrayWithDefaultValues,
  createIdList,
  getRandomEntity,
  createCardList,
} from 'Redux/helpers'

import { selectors } from 'Redux/Store'
import { ExpeditionsStateSlice } from '../types'
import { TreasureIdsStateSlice } from 'Redux/Store/Settings/Expansions/Treasures/ids'
import { TreasureContentStateSlice } from 'Redux/Store/Settings/Expansions/Treasures/content'
import { MagesContentStateSlice } from 'Redux/Store/Settings/Expansions/Mages/content'
import { CardsContentStateSlice } from 'Redux/Store/Settings/Expansions/Cards/content'
import { SelectedLanguagesStateSlice } from 'Redux/Store/Settings/Expansions/Languages'

export const rollNewEntity = (
  list: string[],
  getEntity: types.SeededEntityGetter = getRandomEntity,
  seed: types.Seed
): { result: string; seed: types.Seed } => {
  const idList = createIdList(
    list,
    createArrayWithDefaultValues(1, 'EMPTY'),
    getEntity,
    seed
  )
  return {
    result: idList.result[0],
    seed: idList.seed,
  }
}

type RewardResult = {
  result: string[]
  seed: types.Seed
}

// TODO add tests
export const getSupplyIds = ({
  supply,
  seed,
  stillAvailableCardsByType,
}: {
  supply: types.SupplyRewardConfig | undefined
  seed: types.Seed
  stillAvailableCardsByType: {
    Gem: types.ICard[]
    Relic: types.ICard[]
    Spell: types.ICard[]
    EMPTY: never[]
    ANY: never[]
  }
}) => {
  const baseResult = {
    result: [],
    seed,
  }

  if (supply === undefined) {
    return baseResult
  } else {
    // We split blueprints and ids.
    // This has two benefits:
    // 1.) it lets us filter available cards by those ids, which have been explicitely specified
    // 2.) it makes the reduce call a bit easer to read and maintain, as we don't have to handle the 'string' case
    const ids = supply.ids.filter(
      (idOrBlueprint) => typeof idOrBlueprint === 'string'
    )
    const bluePrints = supply.ids.filter(
      (idOrBlueprint): idOrBlueprint is types.IBluePrint =>
        typeof idOrBlueprint !== 'string'
    )

    return bluePrints.reduce(
      (acc: RewardResult, blueprint) => {
        const cardType = blueprint.type
        const stillAvailableCards = stillAvailableCardsByType[cardType].filter(
          (card) => ids.indexOf(card.id) === -1
        )

        const cardCreationResult = createCardList(
          stillAvailableCards,
          [blueprint],
          undefined,
          acc.seed
        )

        return {
          result: [
            ...acc.result,
            ...cardCreationResult.result.map((card) => card.id),
          ],
          seed: cardCreationResult.seed,
        }
      },
      {
        ...baseResult,
        result: ids,
      } as RewardResult // FIXME casting should technically not be necessary (but i could make it work without)
    )
  }
}

export const getTreasureIds = ({
  treasure,
  stillAvailableTreasureIdsByLevel,
  seed,
}: {
  treasure: types.TreasureRewardConfig | undefined
  seed: types.Seed
  stillAvailableTreasureIdsByLevel: {
    1: string[]
    2: string[]
    3: string[]
  }
}) => {
  const baseResult = {
    result: [],
    seed,
  }

  if (treasure === undefined) {
    return baseResult
  } else {
    const ids = treasure.ids.filter(
      (idOrRandom) => typeof idOrRandom === 'string'
    )

    const randomTreasureConfigs = treasure.ids.filter(
      (idOrRandom): idOrRandom is { random: true; level: 1 | 2 | 3 } =>
        typeof idOrRandom !== 'string'
    )

    return randomTreasureConfigs.reduce(
      (acc: RewardResult, config) => {
        const treasureLevel = config.level
        const stillAvailableCards = stillAvailableTreasureIdsByLevel[
          treasureLevel
        ].filter((treasureId) => ids.indexOf(treasureId) === -1)

        const treasureIdsResult = createIdList(
          stillAvailableCards,
          ['EMPTY'],
          undefined,
          acc.seed
        )

        return {
          result: [...acc.result, ...treasureIdsResult.result],
          seed: treasureIdsResult.seed,
        }
      },
      {
        ...baseResult,
        result: ids,
      } as RewardResult // FIXME casting should technically not be necessary (but i could make it work without)
    )
  }
}

export const getMageIds = ({
  mage,
  stillAvailableMageIds,
  seed,
}: {
  mage: types.MageRewardConfig | undefined
  seed: types.Seed
  stillAvailableMageIds: string[]
}) => {
  const baseResult = {
    result: [],
    seed,
  }

  if (mage === undefined) {
    return baseResult
  } else {
    const ids = mage.ids.filter((idOrRandom) => typeof idOrRandom === 'string')

    const randomTreasureConfigs = mage.ids.filter(
      (idOrRandom): idOrRandom is { random: true } =>
        typeof idOrRandom !== 'string'
    )

    return randomTreasureConfigs.reduce(
      (acc: RewardResult, _) => {
        const filteredMageIds = stillAvailableMageIds.filter(
          (mageId) => ids.indexOf(mageId) === -1
        )

        const mageIdsResult = createIdList(
          filteredMageIds,
          ['EMPTY'],
          undefined,
          acc.seed
        )

        return {
          result: [...acc.result, ...mageIdsResult.result],
          seed: mageIdsResult.seed,
        }
      },
      {
        ...baseResult,
        result: ids,
      } as RewardResult // FIXME casting should technically not be necessary (but i could make it work without)
    )
  }
}

// TODO add tests
export const handleCustomRewards = (
  state: ExpeditionsStateSlice &
    CardsContentStateSlice &
    TreasureContentStateSlice &
    TreasureIdsStateSlice &
    MagesContentStateSlice &
    SelectedLanguagesStateSlice,
  rewardConfig: {
    type: 'custom'
    treasure?: types.TreasureRewardConfig
    mage?: types.MageRewardConfig
    supply?: types.SupplyRewardConfig
  },
  expedition: types.Expedition
) => {
  const expeditionId = expedition.id

  ///////////////////////
  // Content Selection //
  ///////////////////////

  const stillAvailableCardsByType = {
    Gem: selectors.getStillAvailableGems(state, {
      expeditionId,
    }),
    Relic: selectors.getStillAvailableRelics(state, {
      expeditionId,
    }),
    Spell: selectors.getStillAvailableSpells(state, {
      expeditionId,
    }),
    EMPTY: [],
    ANY: [],
  }

  const stillAvailableTreasureIdsByLevel = {
    1: selectors.getStillAvailableTreasureIdsByLevel(state, {
      treasureLevel: 1,
      expeditionId,
    }),
    2: selectors.getStillAvailableTreasureIdsByLevel(state, {
      treasureLevel: 2,
      expeditionId,
    }),
    3: selectors.getStillAvailableTreasureIdsByLevel(state, {
      treasureLevel: 3,
      expeditionId,
    }),
  }

  const stillAvailableMageIds = selectors.getStillAvailableMageIds(state, {
    expeditionId,
  })

  ////////////
  // Supply //
  ////////////

  const { supply } = rewardConfig
  const supplyIdsResult = getSupplyIds({
    supply,
    seed: {
      seed: expedition.seed.seed,
      state: expedition.seed.supplyState,
    },
    stillAvailableCardsByType,
  })

  ///////////
  // Mages //
  ///////////

  const { mage } = rewardConfig
  const mageIdsResult = getMageIds({
    mage,
    seed: supplyIdsResult.seed,
    stillAvailableMageIds,
  })

  ///////////////
  // Treasures //
  ///////////////

  const { treasure } = rewardConfig
  const treasureIdsResult = getTreasureIds({
    treasure,
    seed: mageIdsResult.seed,
    stillAvailableTreasureIdsByLevel,
  })

  return {
    rewards: {
      supplyIds: supplyIdsResult.result,
      treasure: treasureIdsResult.result,
      mages: mageIdsResult.result,
    },
    seed: treasureIdsResult.seed,
  }
}
