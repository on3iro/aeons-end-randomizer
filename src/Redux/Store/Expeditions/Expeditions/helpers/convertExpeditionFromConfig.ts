import * as types from 'aer-types/types'
import shortid from 'shortid'
import { getLatestMigrationVersion } from 'Redux/Store/Expeditions/Expeditions/migrations'

export const convertExpeditionFromConfig = (
  config: types.ImportedExpeditionConfig
): types.ExpeditionFromImportedConfig => {
  const expeditionId = shortid.generate()

  const expedition = {
    id: expeditionId,
    name: config.name,
    bigPocketVariant: config.bigPocketVariantConfig,
    uniqueMageNames: config.uniqueMageNamesConfig || false,
    score: 0,
    seed: {
      seed: config.seedConfig || shortid.generate(),
      supplyState: true,
      nemesisState: true,
    },
    barracks: {
      supplyIds: config.initialBarracksConfig?.supplyIds || [],
      mageIds: config.initialBarracksConfig?.mageIds || [],
      treasureIds: config.initialBarracksConfig?.treasureIds || [],
    },
    banished: [],
    finished: false,
    upgradedBasicNemesisCards: config.initialUBNCardsConfig || [],
    settingsSnapshot: config.settingsSnapshotConfig,
    sequence: {
      firstBranchId: config.sequenceConfig.firstBranchId,
      branches: Object.keys(config.sequenceConfig.branches)
        .map((key) => {
          const branch = {
            ...config.sequenceConfig.branches[key],
            id: key,
            expeditionId,
            status:
              key === config.sequenceConfig.firstBranchId
                ? 'unlocked'
                : 'locked',
            ...(config.sequenceConfig.branches[key].type === 'battle'
              ? { tries: 0 }
              : {}),
          }

          return branch
        })
        .reduce(
          (acc, branch) => ({
            ...acc,
            [branch.id]: branch,
          }),
          {}
        ),
    },
    migrationVersion: getLatestMigrationVersion(),
  }

  return expedition
}
