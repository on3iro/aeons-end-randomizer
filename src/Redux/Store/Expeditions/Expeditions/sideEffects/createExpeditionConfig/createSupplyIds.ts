import { byCost } from 'helpers'

import { createSupply } from 'Redux/helpers'
import * as types from 'aer-types/types'

export const createSupplyIds = (
  availableCards: types.ICard[],
  supplySetup: types.IMarketSetup,
  seed: types.Seed
) => {
  const { gems, relics, spells, seed: resultSeed } = createSupply(
    availableCards,
    supplySetup.tiles,
    seed
  )
  const gemsByCost = gems.sort(byCost)
  const relicsByCost = relics.sort(byCost)
  const spellsByCost = spells.sort(byCost)

  return {
    result: [...gemsByCost, ...relicsByCost, ...spellsByCost].map(
      (card) => card.id
    ),
    seed: resultSeed,
  }
}
