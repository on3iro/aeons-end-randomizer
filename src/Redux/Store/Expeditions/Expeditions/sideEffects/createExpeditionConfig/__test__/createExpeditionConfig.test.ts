import shortidMock from 'shortid'

jest.mock('shortid', () => {
  return {
    generate: () => 'TEST_EXPEDITION_ID',
  }
})

import { createExpedition } from '../index'

import * as helpersModule from 'Redux/helpers'
import * as createSupplyIdsModule from '../createSupplyIds'
import * as createTreasureIdsModule from '../createTreasureIds'
import { rootState } from '__fixtures__/rootState'

describe('createExpeditionConfig()', () => {
  const getState = () => rootState

  afterAll(() => {
    jest.restoreAllMocks()
  })

  it('should produce correct result', () => {
    const result = createExpedition(getState, {
      variantId: 'SHORT',
      name: 'Test expedition config',
      bigPocketVariant: true,
      uniqueMageNames: false,
      marketId: 'market1',
      seedValue: 'TEST',
    })

    expect(result).toMatchSnapshot()
  })

  it('should use provided seedValue', () => {
    const createIdListSpy = jest.spyOn(helpersModule, 'createIdList')

    createExpedition(getState, {
      variantId: 'SHORT',
      name: 'Test expedition config',
      bigPocketVariant: true,
      uniqueMageNames: false,
      marketId: 'market1',
      seedValue: 'TEST_SEED',
    })

    expect(createIdListSpy.mock.calls[0]).toEqual([
      expect.any(Array),
      expect.any(Array),
      expect.any(Function),
      {
        seed: 'TEST_SEED',
      },
    ])

    createIdListSpy.mockRestore()
  })

  it('should fallback to expeditionId as seed value', () => {
    const createIdListSpy = jest.spyOn(helpersModule, 'createIdList')

    createExpedition(getState, {
      variantId: 'SHORT',
      name: 'Test expedition config',
      bigPocketVariant: true,
      uniqueMageNames: false,
      marketId: 'market1',
    })

    expect(createIdListSpy.mock.calls[0]).toEqual([
      expect.any(Array),
      expect.any(Array),
      expect.any(Function),
      {
        seed: 'TEST_EXPEDITION_ID',
      },
    ])

    createIdListSpy.mockRestore()
  })

  it('should use intermediate seed states', () => {
    const createIdListSpy = jest.spyOn(helpersModule, 'createIdList')
    const createSupplyIdsSpy = jest.spyOn(
      createSupplyIdsModule,
      'createSupplyIds'
    )
    const createTreasureIdsSpy = jest.spyOn(
      createTreasureIdsModule,
      'createTreasureIds'
    )

    const result = createExpedition(getState, {
      variantId: 'SHORT',
      name: 'Test expedition config',
      bigPocketVariant: true,
      marketId: 'market1',
    })

    expect(createIdListSpy.mock.calls[0]).toEqual([
      expect.any(Array),
      expect.any(Array),
      expect.any(Function),
      {
        seed: 'TEST_EXPEDITION_ID',
      },
    ])

    expect(createSupplyIdsSpy.mock.calls[0]).toEqual([
      expect.any(Array),
      expect.any(Object),
      createIdListSpy.mock.results[0].value.seed,
    ])

    expect(createTreasureIdsSpy.mock.calls[0]).toEqual([
      expect.any(Object),
      expect.any(Array),
      createSupplyIdsSpy.mock.results[0].value.seed,
    ])

    expect(result.seed).toEqual({
      seed: createTreasureIdsSpy.mock.results[0].value.seed.seed,
      supplyState: createTreasureIdsSpy.mock.results[0].value.seed.state,
      nemesisState: true,
    })

    createIdListSpy.mockRestore()
    createSupplyIdsSpy.mockRestore()
    createTreasureIdsSpy.mockRestore()
  })
})
