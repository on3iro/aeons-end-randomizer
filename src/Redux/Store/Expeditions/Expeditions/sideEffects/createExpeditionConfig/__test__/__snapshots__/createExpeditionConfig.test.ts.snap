// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createExpeditionConfig() should produce correct result 1`] = `
Object {
  "banished": Array [],
  "barracks": Object {
    "mageIds": Array [
      "Ulgimor",
      "Reeve",
      "Mazahaedron",
      "LegacyGreen",
    ],
    "supplyIds": Array [
      "VriswoodAmber",
      "JeweledBrain",
      "BreachOre",
      "GalvanizedBauble",
      "CairnCompass",
      "ThermalDart",
      "ForceTransfusion",
      "Equilibrium",
      "ReduceToAsh",
    ],
    "treasureIds": Array [],
  },
  "bigPocketVariant": true,
  "finished": false,
  "id": "TEST_EXPEDITION_ID",
  "migrationVersion": 2020030401,
  "name": "Test expedition config",
  "score": 0,
  "seed": Object {
    "nemesisState": true,
    "seed": "TEST",
    "supplyState": Object {
      "S": Array [
        219,
        164,
        228,
        57,
        11,
        186,
        127,
        134,
        28,
        167,
        15,
        223,
        45,
        213,
        242,
        109,
        72,
        185,
        240,
        14,
        118,
        34,
        106,
        88,
        211,
        23,
        135,
        49,
        188,
        65,
        174,
        54,
        16,
        39,
        172,
        129,
        165,
        162,
        75,
        52,
        189,
        122,
        59,
        103,
        58,
        202,
        77,
        197,
        50,
        113,
        155,
        210,
        225,
        71,
        100,
        154,
        115,
        56,
        224,
        168,
        91,
        40,
        85,
        0,
        160,
        247,
        73,
        24,
        207,
        31,
        150,
        83,
        181,
        138,
        37,
        98,
        236,
        6,
        3,
        241,
        35,
        2,
        93,
        177,
        141,
        26,
        184,
        201,
        69,
        112,
        114,
        8,
        231,
        171,
        76,
        130,
        27,
        151,
        146,
        214,
        107,
        1,
        124,
        148,
        203,
        55,
        229,
        10,
        255,
        137,
        53,
        173,
        89,
        101,
        120,
        218,
        144,
        194,
        175,
        178,
        126,
        209,
        208,
        149,
        4,
        117,
        179,
        159,
        99,
        230,
        169,
        36,
        156,
        87,
        22,
        33,
        193,
        200,
        217,
        119,
        250,
        190,
        253,
        249,
        62,
        131,
        157,
        199,
        68,
        30,
        170,
        143,
        238,
        41,
        60,
        227,
        161,
        145,
        7,
        251,
        198,
        123,
        235,
        86,
        142,
        21,
        192,
        95,
        25,
        61,
        132,
        158,
        67,
        63,
        105,
        136,
        92,
        139,
        102,
        97,
        204,
        18,
        78,
        43,
        246,
        195,
        234,
        38,
        128,
        66,
        81,
        243,
        183,
        108,
        74,
        226,
        221,
        216,
        19,
        166,
        180,
        79,
        17,
        42,
        182,
        187,
        222,
        248,
        111,
        47,
        82,
        13,
        32,
        46,
        133,
        254,
        48,
        232,
        215,
        12,
        191,
        252,
        20,
        90,
        94,
        110,
        80,
        5,
        70,
        196,
        44,
        140,
        121,
        205,
        239,
        212,
        125,
        176,
        29,
        84,
        244,
        237,
        233,
        206,
        153,
        51,
        116,
        147,
        220,
        245,
        96,
        152,
        64,
        104,
        9,
        163,
      ],
      "i": 91,
      "j": 16,
    },
  },
  "sequence": Object {
    "branches": Object {
      "TEST_EXPEDITION_ID": Object {
        "config": Object {
          "tier": Object {
            "isNewTier": true,
            "tier": 4,
          },
          "treasure": Object {
            "hasTreasure": false,
          },
        },
        "expeditionId": "TEST_EXPEDITION_ID",
        "id": "TEST_EXPEDITION_ID",
        "nextBranchId": undefined,
        "status": "locked",
        "tries": 0,
        "type": "battle",
      },
    },
    "firstBranchId": "TEST_EXPEDITION_ID",
  },
  "settingsSnapshot": Object {
    "availableCardIds": Array [
      "FeralLightning",
      "PlanarInsight",
      "SpectralEcho",
      "ConsumingVoid",
      "CloudedSapphire",
      "LavaTendril",
      "BottledVortex",
      "ArcaneNexus",
      "DarkFire",
      "PhoenixFlame",
      "Jade",
      "AmplifyVision",
      "VriswoodAmber",
      "BlastingStaff",
      "SiftersPearl",
      "WildfireWhip",
      "FocusingOrb",
      "OblivionSwell",
      "TripliteCore",
      "NerveJab",
      "DoubleTap",
      "ObliviumResin",
      "Summonite",
      "ManifoldContainer",
      "DrainingTouch",
      "JoltingCrust",
      "MentiteChunk",
      "SoulCords",
      "VoidiumSpike",
      "PrecisionShot",
      "MutedLacosite",
      "BladeOfWisdom",
      "StunningForce",
      "FlashOfIntellect",
      "ScholarsOpus",
      "BottledStar",
      "BurningCurrent",
      "PsychicEruption",
      "LivingGauntlet",
      "ShiningTetrite",
      "VoidBond",
      "Combustion",
      "VimDynamo",
      "DisintegratingScythe",
      "MonstrousInferno",
      "DevouringShadow",
      "BanishingTopaz",
      "Transmogrifier",
      "JeweledBrain",
      "BlastSphere",
      "WoundMender",
      "BranchingRadite",
      "FireChakram",
      "BraneKnife",
      "FulmiteSlab",
      "GildedMarble",
      "VoltaicRelay",
      "SparkingSiphon",
      "BreachCommunion",
      "InfernalMirror",
      "Geophage",
      "ForceTransfusion",
      "ConductiveGrit",
      "AncientCyanolith",
      "EtherealHand",
      "BendingBeam",
      "EntangledShard",
      "ArcaneSalvo",
      "PhasedPortalite",
      "PyroGeist",
      "BreachFlare",
      "ThunderousOath",
      "SoothingTorporene",
      "ArcingSilicate",
      "WarpingHaze",
      "IncineratingFist",
      "RealityStabilizer",
      "BreachCollision",
      "FieryConclusion",
      "StarfireFrenzy",
      "CrumblingCompound",
      "PropheticLens",
      "AdrenalBatteries",
      "RefinedLumenium",
      "VoidMill",
      "GravityNode",
      "NeuralWreath",
      "SphereOfInversion",
      "LurkingSionite",
      "UnrefinedBlaststone",
      "Exogranite",
      "RecurringJasper",
      "EnergizedRubidium",
      "ResonantPearl",
      "CagedFire",
      "GalvanizedBauble",
      "AetherConduit",
      "WellOfEnergy",
      "PainConduit",
      "LinkConduit",
      "MarbleGalaxy",
      "DelugeOfPower",
      "ReverberatingShock",
      "FatalHarmony",
      "BouncingBoom",
      "SpiritLift",
      "Tethered Darts",
      "ForceAmplifier",
      "ImbuedSmash",
      "PatternStrike",
      "MoltenHammer",
      "Radiance",
      "Blaze",
      "ScryingBolt",
      "SagesBrand",
      "LeechingAgate",
      "TemporalHelix",
      "Char",
      "AlienElement",
      "Scorch",
      "Pyromancy",
      "FeedbackAura",
      "Catalyst",
      "PainStone",
      "AstralCube",
      "RiddleSphere",
      "NetherConduit",
      "HauntedBerylite",
      "Olivinite",
      "CacheGlass",
      "ReflectiveConduit",
      "WillWeaver",
      "BreachSeeker",
      "CrescendoRay",
      "StormVapors",
      "MantraOfStrength",
      "EmbodyFlame",
      "SnapRitual",
      "ShiningFluorite",
      "PolyphaseTurbine",
      "BreachExtractor",
      "FocusingConduit",
      "DualFlash",
      "DustCaller",
      "FossilizedScarab",
      "ThermalDart",
      "Resonate",
      "Conflagration",
      "DimensionalKey",
      "InnerFire",
      "Fulminate",
      "EternityCharm",
      "ErraticIngot",
      "ReduceToAsh",
      "VortexGauntlet",
      "Carbonize",
      "Equilibrium",
      "VolcanicGlass",
      "ConvectionField",
      "JaggedLightning",
      "BloodstoneJewel",
      "Kindle",
      "ThoughtformFamiliar",
      "Crystallize",
      "FieryTorrent",
      "CelestialSpire",
      "ConjureTheLost",
      "NovaForge",
      "DreadDiamond",
      "CairnCompass",
      "ScoriaSlag",
      "BreachOre",
      "FiendCatcher",
      "PyrotechnicSurge",
      "Aurora",
      "ConclaveScroll",
      "PrimordialFetish",
      "FrozenMagmite",
      "MagesTotem",
      "DrownInFlames",
      "FleetingVision",
      "SplinterMissile",
      "EchoStone",
      "ThievingSpirit",
      "FracturedQuartz",
      "RockLauncher",
      "AbacusOfIgnition",
      "IlluminatingFlame",
      "ChronalArc",
      "SymbioticSynapse",
    ],
    "availableMageIds": Array [
      "Kadir",
      "Lash",
      "Mist",
      "Phaedraxa",
      "Nym",
      "Reeve",
      "Zhana",
      "Inco",
      "Razra",
      "LegacyGreen",
      "LegacyYellow",
      "LegacyOrange",
      "LegacyBlue",
      "Sahala",
      "Gygar",
      "Soskel",
      "Taqren",
      "Talix",
      "Rhia",
      "Claudia",
      "Lost",
      "Malastar",
      "Indira",
      "Remnant",
      "Nook",
      "Mazra",
      "Qu",
      "Sparrow",
      "XaxosTV",
      "Dezmodia",
      "Garu",
      "Gex",
      "Mazahaedron",
      "MistWE",
      "Quilius",
      "Ulgimor",
      "YanMagda",
      "Jian",
    ],
    "availableNemesisIds": Array [
      "Rageborne",
      "HordeCrone",
      "Necroswarm",
      "Bladius",
      "Deathmind",
      "FungalMesh",
      "MaelstromRisen",
      "SpawningHorror",
      "XaxosAscended",
      "Maggoth",
      "Arachnos",
      "AgelessWalker",
      "Fenrix",
      "BlightLord",
      "WaywardOne",
      "ThriceDeadProphet",
      "Wraithmonger",
      "TheWailing",
      "TheWanderer",
      "KnightOfShackles",
      "MaidenOfThorns",
      "GateWitch",
      "HollowCrown",
      "MagusOfCloaks",
      "UmbraTitan",
      "PrinceOfGluttons",
      "CarapaceQueen",
      "CrookedMask",
    ],
    "availableTreasureIds": Array [
      "DezmodiasOblivionShard",
      "UlgimorsCoalFragment",
      "IndirasLoneOpal",
      "MistsAmethystParagon",
      "MalastarsImmolate",
      "YanMagdasIlluminate",
      "EyeOfTheMaelstrom",
      "WellOfDespair",
      "CronesAmulet",
      "EdibleFungusChunks",
      "UmbralHornOfWar",
      "TheBrokenMask",
      "AlchemistsAlembic",
      "SoothsayersPouch",
      "SuppressingPin",
      "Quicksilver",
      "BandOfRetrieval",
      "PurifiedBangle",
      "KadirsEmeraldShard",
      "GexsShatteredGeode",
      "JiansMoonstoneShard",
      "LashsQuartzShard",
      "XaxosFlare",
      "AdelheimsAmethystShard",
      "BramasBuriedLight",
      "GarusTorch",
      "XaxosPyre",
      "MistsGarnetShard",
      "ShimmeringCloakOfTheMagus",
      "ImbuedShackles",
      "GluttonsTooth",
      "CoreOfRage",
      "CleansingAmulet",
      "CarapaceFragement",
      "ThornedWhip",
      "BonesOfDeathmind",
      "AcceleratingGauntlets",
      "EssenceExtractor",
      "BroochOfAttunment",
      "BladedCrystal",
      "PrismOfDestruction",
      "ForgottenTrinket",
      "PrecisionMagnifier",
      "ShroudOfObfucation",
      "VerdantStaff",
      "FlowingMantle",
      "MazahaedronsWorldheartShard",
      "PhaedraxasTourmalineShard",
      "FracturedShell",
      "Extra-DimensionalLens",
      "SeersBracer",
      "RemnantsVoidShard",
      "QuiliussErasure",
      "ZhanasFleetingEmber",
      "BlightedRootClump",
      "PetrifiedWitchFinger",
      "WaywardScraps",
      "TrueSightMonocle",
      "EntwinedAetherStrands",
      "SiphoningBlade",
      "GemEncrustedAnklet",
      "ReevesRavenShard",
      "NymsVision",
      "SparrowsAid",
      "WraithsEssence",
      "VolatileClasp",
      "BreachStabilizer",
      "FangedChoker",
      "EndlessBandolier",
    ],
    "availableUpgradedBasicNemesisCardIds": Array [
      "HissingAcid",
      "Wreck",
      "BaneCommander",
      "SkyTremor",
      "EngulfingMadness",
      "Separate",
      "NeedleDoom",
      "Marked",
      "Eradicate",
      "FlashOfDecay",
      "Terrify",
      "RayOfDesolation",
      "ThunderousTempest",
      "Sacrifice",
      "Burialskulk",
      "Rupture",
      "FurrowOfDestruction",
      "Mangle",
      "HowlingSlayer",
    ],
    "supplySetup": Object {
      "active": false,
      "id": "market1",
      "name": "Market Setup 1",
      "tiles": Array [
        Object {
          "operation": "<",
          "threshold": 4,
          "type": "Gem",
        },
        Object {
          "operation": "=",
          "threshold": 4,
          "type": "Gem",
        },
        Object {
          "operation": "ANY",
          "type": "Gem",
        },
        Object {
          "operation": "ANY",
          "type": "Relic",
        },
        Object {
          "operation": "ANY",
          "type": "Relic",
        },
        Object {
          "operation": "<",
          "threshold": 5,
          "type": "Spell",
        },
        Object {
          "operation": "<",
          "threshold": 5,
          "type": "Spell",
        },
        Object {
          "operation": ">",
          "threshold": 5,
          "type": "Spell",
        },
        Object {
          "operation": ">",
          "threshold": 5,
          "type": "Spell",
        },
      ],
      "type": "official",
    },
    "usedExpansions": Array [
      "AE",
      "BS",
      "Depths",
      "IW",
      "Legacy",
      "NA",
      "Nameless",
      "OD",
      "SD",
      "TA",
      "TV",
      "WE",
      "promos",
    ],
  },
  "uniqueMageNames": false,
  "upgradedBasicNemesisCards": Array [],
}
`;
